libs/core/src/visualization.cpp

```diff
@@ -281,6 +281,52 @@ void FeatureLayerVisualization::addFeature(
             }
             // Iterate over all the layer's attributes.
             layer->forEachAttribute([&, this](auto&& attr){
+                // Check if the attribute's values match the attribute mask for the rule.
+                if (auto const& attrMask = rule.attributeMask()) {
```

Comment:
Just run this inside of `addAttribute` then we can get rid of the simfil context setup duplication.